File: Textfield And Buttons/[Objective-C] textfieldAndButtons/textfieldAndButtons/ViewController.m
//
//  ViewController.m
//  textfieldAndButtons
//
//  Created by Luis Andres Castillo Hernandez on 11/21/15.
//  Copyright © 2015 DevBros. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
{
    
}
@end

@implementation ViewController

@synthesize topBorder, bottomBorder, leftBorder, rightBorder, buttonWithBorder;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    
    [self addTopBorderWithColor:[UIColor blueColor] andWidth:2.0 buttonToUpdate:topBorder];
    
    
    [self addBottomBorderWithColor:[UIColor lightGrayColor] andWidth:2.0 buttonToUpdate:bottomBorder];
    
    
    [self addLeftBorderWithColor:[UIColor redColor] andWidth:2.0 buttonToUpdate:leftBorder];
    
    
    [self addRightBorderWithColor:[UIColor greenColor] andWidth:2.0 buttonToUpdate:rightBorder];
    
        //buttonWithBorder
    [self MakeRoundBorder:buttonWithBorder withColor:[UIColor darkGrayColor] andWidth:0.8 andRadius:10];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



- (void)addTopBorderWithColor:(UIColor *)color
                     andWidth:(CGFloat) borderWidth
               buttonToUpdate:(UIButton *) buttonProvided
{
    CALayer *border = [CALayer layer];
    border.backgroundColor = color.CGColor;
    
    border.frame = CGRectMake(0, 0, buttonProvided.frame.size.width, borderWidth);
    [buttonProvided.layer addSublayer:border];
}//eom

- (void)addBottomBorderWithColor:(UIColor *)color
                        andWidth:(CGFloat) borderWidth
                  buttonToUpdate:(UIButton *) buttonProvided
{
    CALayer *border = [CALayer layer];
    border.backgroundColor = color.CGColor;
    
    border.frame = CGRectMake(0, buttonProvided.frame.size.height - borderWidth, buttonProvided.frame.size.width, borderWidth);
    [buttonProvided.layer addSublayer:border];
}//eom

- (void)addLeftBorderWithColor:(UIColor *)color
                      andWidth:(CGFloat) borderWidth
                buttonToUpdate:(UIButton *) buttonProvided
{
    CALayer *border = [CALayer layer];
    border.backgroundColor = color.CGColor;
    
    border.frame = CGRectMake(0, 0, borderWidth, buttonProvided.frame.size.height);
    [buttonProvided.layer addSublayer:border];
}//eom

- (void)addRightBorderWithColor:(UIColor *)color
                       andWidth:(CGFloat) borderWidth
                 buttonToUpdate:(UIButton *) buttonProvided
{
    CALayer *border = [CALayer layer];
    border.backgroundColor = color.CGColor;
    
    border.frame = CGRectMake(buttonProvided.frame.size.width - borderWidth, 0, borderWidth, buttonProvided.frame.size.height);
    [buttonProvided.layer addSublayer:border];
}//eom


-(void)MakeRoundBorder:(UIButton *) button
                withColor:(UIColor *)color
            andWidth:(CGFloat) borderWidth
              andRadius:(CGFloat) cornerRadius
{

    button.layer.borderWidth = borderWidth;
    button.layer.borderColor = color.CGColor;
    button.layer.cornerRadius = cornerRadius;
}//eom

@end

File: Textfield And Buttons/[Swift] textfieldAndButtons/textfield/ViewController.swift
//
//  ViewController.swift
//  textfield
//
//  Created by Luis Andres Castillo Hernandez on 11/17/15.
//  Copyright © 2015 devbros. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITextFieldDelegate {

    
    @IBOutlet weak var buttonWithBorder: UIButton!
    
    @IBOutlet weak var textfieldWithNoBorder: UITextField!
    @IBOutlet weak var textfieldWithBottomOnly: UITextField!
    //
    
    @IBOutlet weak var autoResizeLabelTextfield: UITextField!
    @IBOutlet weak var AutoResizeLabel: UILabel!
    
    //textfield that changes color
    @IBOutlet weak var textfieldChangesOnEdit: UITextField!
    var borderFORtextfieldChangesOnEdit:CALayer = CALayer()
    
    
    
     //MARK: View loading
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func viewDidAppear(animated: Bool)
    {
        
        self.updateTextFieldWithBottomBorderOnly(textfieldWithBottomOnly);
        
        self.setupTextfieldThatChangesColorOnEdit()
        
        self.updateButtonWithBorder()
    }//eom

     //MARK: Memory
    
    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    //MARK:

    func updateTextFieldWithBottomBorderOnly(textfieldProvided:UITextField)
    {
        let border          = CALayer()
        let width           = CGFloat(1.5)
        border.borderColor  = UIColor.lightGrayColor().CGColor
        border.frame        = CGRect(x: 0, y: textfieldProvided.frame.size.height - width, width:  textfieldProvided.frame.size.width+7, height: textfieldProvided.frame.size.height)
        
        border.borderWidth  = width
        textfieldProvided.layer.addSublayer(border)
        textfieldProvided.layer.masksToBounds = true
        
        /*
        
        //objective-c
            CALayer *border = [CALayer layer];
            CGFloat borderWidth = 2;
            border.borderColor = [UIColor darkGrayColor].CGColor;
            border.frame = CGRectMake(0, textField.frame.size.height - borderWidth, textField.frame.size.width, textField.frame.size.height);
            border.borderWidth = borderWidth;
            [textField.layer addSublayer:border];
            textField.layer.masksToBounds = YES;
        */

    }//eom
    
    
    
    //MARK: textfield that changes color on edit
    
    /* setting up textfield with border */
    func setupTextfieldThatChangesColorOnEdit()
    {
        let width           = CGFloat(1.5)
        borderFORtextfieldChangesOnEdit.borderColor  = UIColor.lightGrayColor().CGColor
        borderFORtextfieldChangesOnEdit.frame        = CGRect(x: 0, y: textfieldChangesOnEdit.frame.size.height - width, width:  textfieldChangesOnEdit.frame.size.width+7, height: textfieldChangesOnEdit.frame.size.height)
        
        borderFORtextfieldChangesOnEdit.borderWidth  = width
        textfieldChangesOnEdit.layer.addSublayer(borderFORtextfieldChangesOnEdit)
        textfieldChangesOnEdit.layer.masksToBounds = true
    }//eom
    
    /* chaning textfield colors */
    @IBAction func textfieldEditModeON(sender: UITextField)
    {
        textfieldChangesOnEdit.textColor = UIColor.blueColor()
        borderFORtextfieldChangesOnEdit.borderColor = UIColor.blueColor().CGColor
    
    }//eoa
    
    /* returning textfield colors back to original colors */
    @IBAction func textfieldEditModeOFF(sender: UITextField)
    {
        textfieldChangesOnEdit.textColor = UIColor.blackColor()
        borderFORtextfieldChangesOnEdit.borderColor = UIColor.lightGrayColor().CGColor
    }//eoa
    
    
    //MARK: 
    
    /* releasing first responder if user clicks on background */
    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        textfieldWithNoBorder .resignFirstResponder()
        textfieldWithBottomOnly .resignFirstResponder()
        textfieldChangesOnEdit .resignFirstResponder()
  
    }//eom
    
    //MARK: textfield delegates
    func textFieldShouldEndEditing(textField: UITextField) -> Bool {
    
        textField .resignFirstResponder()
        
        return true;
    }//eom
    
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        textField .resignFirstResponder()
        
        if textField == autoResizeLabelTextfield
        {
            AutoResizeLabel.text = autoResizeLabelTextfield.text
            autoResizeLabelTextfield.text = ""
        }
        
        return true;
    }//eom
    
    //MARK: Buttons
    func updateButtonWithBorder()
    {
        
        self.buttonWithBorder.layer.borderWidth = 0.8
        self.buttonWithBorder.layer.borderColor = UIColor.redColor().CGColor
        
        //make round border
        self.buttonWithBorder.layer.cornerRadius = 10.0
        
    }//eom
    
}//eoc

